pkg/network/tunnel/api/message.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2023 Cisco and/or its affiliates. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//       https://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package api
+
+import (
+	"encoding/json"
+	"io"
+
+	"github.com/google/uuid"
+)
+
+type Message struct {
+	ID   string      `json:"id"`
+	Type MessageType `json:"type"`
+	Data []byte      `json:"data"`
+}
+
+func (m Message) Decode(v any) error {
+	return json.Unmarshal(m.Data, &v)
+}
+
+type MessageType string
+
+const (
+	OpenTCPStreamMessageType     MessageType = "openTCPStream"
+	OpenControlStreamMessageType MessageType = "openControlStream"
+	AddPortMessageType           MessageType = "addPort"
+	AddPortResponseMessageType   MessageType = "addPortResponse"
+	PongMessageType              MessageType = "pong"
+	PingMessageType              MessageType = "ping"
+	RequestConnectionMessageType MessageType = "requestConnection"
+	StreamOpenedMessageType      MessageType = "streamOpened"
+
+	ErrCtrlStreamAlreadyExistsMessageType MessageType = "errCtrlStreamAlreadyExists"
+	ErrInvalidStreamTypeMessageType       MessageType = "errInvalidStreamType"
+	ErrInvalidStreamIDMessageType         MessageType = "errInvalidStreamID"
+)
+
+type OpenTCPStreamMessage struct {
+	ID string `json:"id"`
+}
+
+type AddPortRequestMessage struct {
+	Type          string `json:"type"`
+	ID            string `json:"id"`
+	RequestedPort int    `json:"port"`
+}
+
+type AddPortResponseMessage struct {
+	Type         string `json:"tcp"`
+	ID           string `json:"id"`
+	AssignedPort int    `json:"port"`
+	Address      string `json:"address"`
+}
+
+type RequestConnectionMessage struct {
+	PortID        string `json:"portID"`
+	Identifier    string `json:"identifier"`
+	RemoteAddress string `json:"remoteAddress"`
+	LocalAddress  string `json:"localAddress"`
+}
+
+func CreateMessage(messageType MessageType, message interface{}) (Message, []byte, error) {
+	return CreateMessageWithID(uuid.New().String(), messageType, message)
+}
+
+func CreateMessageWithID(id string, messageType MessageType, message interface{}) (Message, []byte, error) {
+	msg := Message{
+		ID:   uuid.New().String(),
```

Comment:
```suggestion
func CreateMessageWithID(id string, messageType MessageType, message interface{}) (Message, []byte, error) {
	msg := Message{
		ID:   id,
```
`id` parameter was never used but generated